### PR TITLE
feat(trainer): On-Policy Distillation Entrypoint

### PR DESCRIPTION
**Pull request by kyuds**
Add support for on-policy distillation as a SkyRL entrypoint.

---

> [!NOTE]
> **Medium Risk**
> New training entrypoint changes rollout generation and advantage math and depends on external teacher APIs; misconfigured tokenizers or backends can fail late, though validation and tests limit silent breakage.
> 
> **Overview**
> Adds a first-class **on-policy distillation (OPD)** path via `skyrl.train.entrypoints.main_opd` and `skyrl/train/opd/`, replacing the old example that treated the reference model as the teacher (`main_on_policy_distill.py` is removed).
> 
> The frozen teacher is scored through **Fireworks** or **remote vLLM** (`TeacherLogprobClient` + HTTP logprob APIs), not a co-located training worker—so it never weight-syncs and can be much larger than the student. **`OPDTrainer`** generates per prompt group in parallel and calls the teacher as each group finishes (overlapping teacher latency with remaining rollouts), then applies a per-token reverse-KL term to advantages after the configured estimator (`kl_coef`, optional **`use_task_reward`** for pure vs mixed distillation). New **`trainer.teacher`** / **`trainer.algorithm.opd`** config and **`validate_opd_cfg`** guard incompatible algorithm settings.
> 
> Example scripts now target `main_opd` with `TEACHER_MODEL` deployment ids; a GSM8K gpt-oss Fireworks smoke script and docs (`on_policy_distillation.mdx`) are added. **`tests/train/opd/test_opd.py`** covers clients, config validation, and trainer behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7328f0278bb0d9b07eedde2ebd11450e9da2284b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/content/docs/examples/on_policy_distillation.mdx [hunk 1]
```diff
@@ -0,0 +1,77 @@
+---
+title: "On-Policy Distillation"
+---
+
+On-policy distillation (OPD) trains a student on its own rollouts while a frozen teacher grades every
+token: the per-token reverse KL to the teacher, `log π_student − log π_teacher`, is the advantage. It keeps
+the on-policy character of RL and adds a dense, per-token signal
```
[the remark this answers is not in the document]
Comment (kyuds) [reply docs/content/docs/examples/on_policy_distillation.mdx]: this is the notation in tinker blog as well. Note that the reverse kl is subtracted from the computed advantages for the final advantage. Therefore, `student - teacher` actually becomes adding `teacher - student`.

File: skyrl/train/opd/trainer.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+from skyrl.train.trainer import RayPPOTrainer
+from skyrl.train.utils.trainer_utils import validate_generator_output
+
+
+class OPDTrainer(RayPPOTrainer):
+    """``RayPPOTrainer`` for on-policy distillation. Works with any ``GeneratorInterface``."""
+
+    def __init__(self, *args, teacher_client: TeacherLogprobClient, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.teacher_client = teacher_client
```
[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/opd/trainer.py:49]: implemented on newer commit

File: docs/content/docs/examples/on_policy_distillation.mdx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+still generating, so teacher latency overlaps with generation. The generator is untouched: any
+`GeneratorInterface` works.
+
+## Quickstart
+
+```bash
+export FIREWORKS_API_KEY=<your key>
+
+# Serverless smoke test: gpt-oss-20b student, gpt-oss-120b teacher on Fireworks serverless.
+uv run examples/train/gsm8k/gsm8k_dataset.py --output_dir $HOME/data/gsm8k
```
[the remark this answers is not in the document]
Comment (kyuds) [reply docs/content/docs/examples/on_policy_distillation.mdx:22]: only claude runs should use `--isolated`. For user runs, it doesn't matter. `--isolated` is not a hard requisite.

File: docs/content/docs/examples/on_policy_distillation.mdx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+- `trainer.teacher.backend=fireworks` (the only backend for now): a Fireworks model id or a dedicated deployment id
+  (`accounts/<account>/deployments/<id>`), scored through the completions API with an integer prompt and
+  `echo_last`. Serverless is fine for checking the plumbing; use a dedicated deployment for training,
+  because serverless replicas disagree on logprobs by more than the OPD signal itself and the Qwen3 (2025)
+  family is not served serverless.
+
+Before any model is loaded, the entrypoint sends one short sequence to the teacher
+`trainer.algorithm.opd.self_test_samples` times concurrently and refuses to train if the answers differ
+by more than `trainer.algorithm.opd.self_test_max_abs_diff` nats. This catches replica-dependent serving,
+models that reject echo, and a mismatched tokenizer before the run starts.
```
[the remark this answers is not in the document]
Comment (kyuds) [reply docs/content/docs/examples/on_policy_distillation.mdx]: will leave preflights as a todo actually. closing for now.
